src/Type/Php/FilterInputDynamicReturnTypeExtension.php

```diff
@@ -0,0 +1,87 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Type\Php;
+
+use PhpParser\Node\Arg;
+use PhpParser\Node\Expr\ConstFetch;
+use PhpParser\Node\Expr\FuncCall;
+use PhpParser\Node\Expr\Variable;
+use PHPStan\Analyser\Scope;
+use PHPStan\Reflection\FunctionReflection;
+use PHPStan\Type\Constant\ConstantBooleanType;
+use PHPStan\Type\DynamicFunctionReturnTypeExtension;
+use PHPStan\Type\NullType;
+use PHPStan\Type\Type;
+use PHPStan\Type\TypeCombinator;
+use function count;
+use function strtolower;
+
+class FilterInputDynamicReturnTypeExtension implements DynamicFunctionReturnTypeExtension
+{
+
+	private const INPUT_TYPE_VARIABLE_NAME_MAP = [
+		'INPUT_GET' => '_GET',
+		'INPUT_POST' => '_POST',
+		'INPUT_COOKIE' => '_COOKIE',
+		'INPUT_SERVER' => '_SERVER',
+		'INPUT_ENV' => '_ENV',
+	];
+
+	public function __construct(private FilterFunctionReturnTypeHelper $filterFunctionReturnTypeHelper)
+	{
+	}
+
+	public function isFunctionSupported(FunctionReflection $functionReflection): bool
+	{
+		return strtolower($functionReflection->getName()) === 'filter_input';
+	}
+
+	public function getTypeFromFunctionCall(FunctionReflection $functionReflection, FuncCall $functionCall, Scope $scope): ?Type
+	{
+		if (count($functionCall->getArgs()) < 2) {
+			return null;
+		}
+
+		$inputVariable = $this->determineInputVariable($functionCall->getArgs()[0]);
```

Comment:
What's special about variables and this function? What do you do with the variable? What about other expression types?

Reply:
filter_input() only works with superglobal variables like GET and such. But this is an internal thing. From the outside you don't pass an expression but a constant that defines which superglobal input is used. This is why I have this weird mapping in the class that creates the superglobal variable expression. I hope that makes sense.

Reply:
Ah and of course my helper thingy expects an expression. So that's basically re-using filter_var() (there you have the var :)) internally for filter_input()

Reply:
this has become more or less irrelevant I'd say, at least I don't have this mapping anymore since all superglobals are currently assumed to have the same type anyways. but that simplified things